### PR TITLE
Simple metadata query class

### PR DESCRIPTION
**Pull request by r-b-g-b**
load.py that allows you to query the metadata database structured as implemented in 60b46d0

Maybe this gets a little specific, but if we're going to have a predetermined structure for the metadata files, we could use a coherent way to access stuff from there.

1. Do we want something like this?
2. Should this be in `src/preprocessing/load.py` or somewhere else? I put it there so it could be easily called from preprocessing scripts without having to mess with python path. Presumably it would only be needed during preprocessing. This leaves open the possibility for other folders (`viz`, `features`, etc.) to have their own `load.py`.

### CODE REVIEW

**Comment by sslivkoff** [comment]
A few things

- I think this is a step in the right direction though I'm not sure yet what the use case will be. In the thread for the previous json pull request I mentioned that I think each project's load.py will have to be customized to accommodate it's particular needs for clipping, concatenation, and other things between runs. Also I would rather have a loading function that gives me a numpy array than documents. Given those two things I think the class you propose could serve as a middle layer that utilizes the json data programmatically and returns the json documents, but does not perform the all the data loading steps that one might want.

- For some examples of what I am talking about check out `/auto/k1/storm/python_path/datasets/datasets/attention_cukur2.py` or `/auto/k1/storm/python_path/datasets/datasets/qa/vqa_pilot.py`. For every mri dataset I deal with I make a single module that of loader functions for bold data and stimulus. In the functions I usually include options for things like 1) whether to concatenate runs vs give list of arrays 2) whether to zscore 3) whether to denan 4) whether to mask, etc. Typical things that you might do to your data when you first load it

- Rant: My personal coding preference is organizing things around functions not classes. Instead of having to create (and remember how to use) an object, I would rather just be able to say `load.load_subject('XX', ...)`. In my experience objects are rarely necessary and usually you can craft things more elegantly with a function based approach. I'm not saying objects always suck just feel a need to evangelize the functional approach whenever I get a chance because I think it's better.

File: {{cookiecutter.project_repo_name}}/src/preprocessing/load.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import json
+
+
+class ExperimentDB(object):
```
Comment (fatmai) [inline {{cookiecutter.project_repo_name}}/src/preprocessing/load.py]: I have a little of an issue with the name "DB" it is confusing.

File: {{cookiecutter.project_repo_name}}/src/preprocessing/load.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        db = self.db if db is None else db
+        results = {'study': {}}
+        for study_name, study in db['study'].iteritems():
+            for subj in study:
+                if query_subject == subj['subject']:
+                    results['study'][study_name] = [subj]
+        return results
+
+    def query_study(self, query_study, db=None):
+        db = self.db if db is None else db
```
Comment (fatmai) [inline {{cookiecutter.project_repo_name}}/src/preprocessing/load.py]: can we add a comment to this. I mean docstring

File: {{cookiecutter.project_repo_name}}/src/preprocessing/load.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    """
+    def __init__(self, experiment_db_path):
+
+        with open(experiment_db_path, 'r') as f:
+            db = json.load(f)
+
+        self.db = db
+
+    def query(self, subject=None, study=None, db=None):
+        db = self.db if db is None else db
```
Comment (fatmai) [inline {{cookiecutter.project_repo_name}}/src/preprocessing/load.py]: docstring missing

**Comment by fatmai** [comment]
about @sslivkoff comments:

- I think load.py here is not to load the data but the metadata. Then you would use this metadata to pass it to your function. I think load.py, the name, is a little misleading. So, I think we should change this to something more sensible.

- Your examples are useful! Basically, we are trying to take first part of your scripts where you define templates, subjects, etc. out of your loading script.

- Object vs. function is a different battle ;) lol

**Comment by r-b-g-b** [comment]
Yeah agree about naming. Ehhh `query.py`? `experiment_metadata.py`? No real preferences for classes or functions. I used class here because it allows you to bake in the path to the metadata file, which is the 99+% use case. The functional way would either require specifying the path for every call or using `functools.partial` which is basically making a class.
```
import experiment_metadata
results = experiment_metadata.query(subject='ML')
```

functional version would be?
```
from query import query_subject
results = experiment_metadata.query('../../data/preprocess_data.json', subject='ML')
```
Open to other suggestions

**Comment by sslivkoff** [comment]
oh ok I think I understand the purpose a little better now

I guess now my question would be what is the advantage of using this code instead of just loading the entire experiment metadata? once you `json.load()` the metadata file don't you already have a metadata datastructure that is appropriate for all metadata needs?

**Comment by sslivkoff** [comment]
fuck didnt mean to close how do I unclose

**Comment by r-b-g-b** [comment]
Yeah, the idea is that this does most of what you'd want to do if you started loading the metadata. 1. Get all the entries for one subject, 2. get all the entries for one study. Can be extended to add less common goals/things I didn't think of.

**Comment by sslivkoff** [comment]
Ok I see. Personally I would rather just tweak the metadata object to make it able to do those things naturally. Looking at the current json metadata, instead of having subjects as a list of dicts as [subject_metadata], use a dict of the form {subject_name: subject_metadata}. Then you can select by subject easily. And by study_name is already an organized categorization. Not sure why 'study' itself is a top level key, what would be some examples of other keys that are on the same level as 'study'?

**Comment by fatmai** [comment]
{subject_name: subject_metadata} was one consideration that we had. It's more a style reason. This way for each study (e.g. stories, movies, etc.) there is a list of subject metadata. The alternative is something like this, 

```

{"study_name": {"S1":  { "surface": "S1fs",
                                    "dbname": "docdb-user",
                                    "xfmname": "YYYYMMDDS1_transform",
                                    "generated_by_name": "DetrendSGolay",
                                    "maskname": "thick",
                                    "sessions": ["YYYYMMDDS1-1",
                                                         "YYYYMMDDS1-2"]}
                          }
}
```

but I think this gets a little ugly.

**Comment by fatmai** [comment]
I made some changes on naming. Also I removed the the last line b/c that didnt do anything, or I didnt get it. Instead init is with a given path. check it out and make changes as you wish.

**Comment by fatmai** [comment]
@r-b-g-b: I updated the query functions. If you could review that part, that would be great. 

`query_subject` returns now a list of subject metadata from different studies. In order to keep track of which study is which list entry, I entered a `study_name` entry and also a `subject` entry to the returned data.

**Comment by r-b-g-b** [comment]
I envisioned the last line being used like this...
```
from load_metadata import experiment_metadata
experiment_metadata.query_subject('S1')
```
It makes almost no difference in this case, BUT in general, I'd prefer the class to be as general as possible. Like if we gave someone the class with our very specific default path, they'd be like "what the heck is that path?" Instead we hard-code our path in the call to instantiate the class. Saves a line of code when importing and querying too.

**Comment by fatmai** [comment]
hmm, but the cookiecutter is designed to have that specific default path.

**Comment by r-b-g-b** [comment]
It might be kind of confusing that `query_subject` and `query_study` return their results in different formats. Originally I had both return results in the same format as the original datastructure (which is now ```{study1: {S1: {stuff}}```. It makes it a bit more verbose to deal with, but allows you to chain the query functions together...

```
# selects S1 from study1
results = experiment_metadata.query_subject('S1')
results = experiment_metadata.query_study('study1', results)
```

That's what the `.query` method used to do, but it relies on keeping the same structure. Of course the downside is that using the `results` becomes a bit more verbose, one more index step than is necessary, i.e.

```
results['S1']['study1']
```
At the very least, we should have both query functions return stuff in the same format as each other (currently one returns list and the other returns dict). Maybe have both return dicts. `query_subject` returns a dict with keys of `study` and  `query_study` returns a dict with keys of `subject`.

**Comment by fatmai** [comment]
I like that idea. I didn't think about it. I would be fine with that. 

`results['S1']['study1']` is ok. in the very first version we had `results['subject']['S1']` etc.

**Comment by fatmai** [comment]
ok I made change to return a dict, do you want to have a look @r-b-g-b. is that doing what you are saying.

**Comment by fatmai** [comment]
can you send an example for `query` usage?

**Comment by fatmai** [comment]
@r-b-g-b, when you move `load_metadata.py` instantiating `from load_metadata import experiment_metadata` fails I think b/c of path information.

**Comment by r-b-g-b** [comment]
You'll have to `pip -e` install or `python setup.py develop` the python module first. Then you'll be able to do:
```
from <<project_name>>.load_metadata import experiment_metadata
```

**Comment by fatmai** [comment]
are you sure? it stills fail for me.

**Comment by r-b-g-b** [comment]
Ah you're right, now the path has changed relative to the metadata file...

**Comment by fatmai** [comment]
@sslivkoff if are you fine with these changes, I will merge this.

**Comment by sslivkoff** [comment]
This is looking good. I like the new metadata datastructure layout.

Right now the class/methods seem very specific to the `preprocessed_fmri.json` file. Will machinery to load the other json files also go in this `ExperimentMetadata` class?

I am fine with including the query functions. Personally though I don't think I would ever use them because they seem like overkill and I prefer simple things that I do not have to remember how to use. Can we include a function something like this:

```
def load_metadata(path=metadata_path):
    with open(metadata_path, 'r') as f:
        return json.load(f)
```

and `ExperimentMetadata.__init__()` could use it to load the file?

**Comment by fatmai** [comment]
cool.

We may expand this function or add new ones for the other json files. I think we will see what makes sense while we proceed.

`__ini__()` alreay loads the json file. So, what you need to do is

```
from load_metadata import ExperimentMetadata
experiment = ExperimentMetadata(experiment_metadata_path=path)
print experiment.metadata
```

**Comment by sslivkoff** [comment]
ah I didnt see that the file was now called `load_metadata.py`, it makes sense why it only loads the metadata!

yes but I would like to decouple the loading from the class, so that users do not have to use the class if they do not want to. I would like to be able to do 

```
import load_metadata
metadata = load_metadata.load_metadata()
```

**Comment by fatmai** [comment]
this will do the same thing, though... I understand not wanting to use the Object methods etc. 

But you could really just add the two lines to your code. This will be the exact same output and you don't have to think/use the object further. We can add an example into the docstring for the user. I just dont see the reason to duplicate the same functionality.

**Comment by sslivkoff** [comment]
right now the code is very opinionated over how to load/use your metadata. I think the cookiecutter should be less unopinionated to appeal to everyone in lab

some people do not want yet another object class in their life and they already know how to use simple nested dicts that do the job just fine

it's not a duplication if `ExperimentMetadata.__init__()` method uses the `load_metadata()` function

**Comment by fatmai** [comment]
I am not necessarily in favor of one or the other one. The object oriented way is compact and easy enough in its current form that I think there is no need to make this a fight OO vs Functions. I agree that the cookiecutter should appeal to everyone, but there are simple decisions that we need to make. If it does not hurt too much that should be fine in my opinion.

do you want to have a function that is outside of the class that the init file uses to load the function? I think that is from a programming viewpoint not very clean. but I am open to suggestions. 

also @r-b-g-b, any opinions on that?

**Comment by sslivkoff** [comment]
I dont really know how to objectively evaluate cleaness but a separate function outside of the class seems clean to me.

But instead of having a function at all what about just having a top level block of code (outside of functions/classes/methods) that does

```
with open(metadata_path, 'r') as f:
    metadata = json.load(f)
```

then metadata exists in the module's namespace and no extra machinery is necessary

**Comment by fatmai** [comment]
@sslivkoff and I talked about this last comment here over beer last week and came up with a solution that seems to work 'elegantly'. @r-b-g-b and @sslivkoff why dont you check this? basically, the loading the json is out of `load_metadata.py` (as we will have different metadata json files and it will be used by other classess as well) and in `load.py`. This way someone who just wants to have a direct access to the json file can use it through this and load.py can in general have bunch of loading wrappers.

**Comment by r-b-g-b** [comment]
I like it!

**Comment by fatmai** [comment]
cool. I will merge this then. @sslivkoff agrees too I think.

**Pull request merged**